### ISSUE TITLE
Use source_lfn and destination_lfn from couch docs

### ISSUE THREAD

**Issue by AndresTanasijczuk** [dmwm/AsyncStageout#4347]
Currently ASO uses the 'lfn' parameter from the Couch documents. This parameter contains the LFN of the source file, i.e. the temporary LFN. For example:
lfn: '/store/temp/user/atanasi.<hash>/<dir1>/<dir2>/<etc>/file.root'
or
lfn: '/store/temp/group/<dir1>/<dir2>/<etc>/file.root'
ASO then deduces the destination LFN from this source LFN by removing the "temp/" (and the ".<hash>" if it exists) to end up with a destination LFN = '/store/user/atanasi/<dir1>/<dir2>/<etc>/file.root' or '/store/group/<dir1>/<dir2>/<etc>/file.root' respectively.
Since some time I have introduced two new parameters in the Couch documents, which are the 'source_lfn' and the 'destination_lfn'. The 'source_lfn' is exactly the same as the 'lfn', so you can safely use 'source_lfn' instead of 'lfn'. That's the first step. And the second step is that ASO should use 'destination_lfn' as the destination LFN for transfers and publication and don't deduce anymore the destination LFN from the source LFN.

**Comment by HassenRiahi** [comment]
It will help also to address this ticket https://github.com/dmwm/AsyncStageout/issues/4334 for non-SE <-> SE transfers

**Comment by HassenRiahi** [comment]
Fixed.

**Comment by AndresTanasijczuk** [comment]
Why is this issue closed? ASO is not using the source_lfn and destination_lfn except for some few places.

**Comment by HassenRiahi** [comment]
Pls could you point out to @dciangot where are the other places in ASO that we are still using lfn instead of source/destination_lfn. Or send anothr patch to have this ticket definitely closed.

**Comment by AndresTanasijczuk** [comment]
These are views where I see that 'doc.lfn' is still used (maybe you don't use most of these views??):
* src/couchapp/AsyncTransfer/views/TransfersByFailuresReasons/map.js
* src/couchapp/AsyncTransfer/views/UserByStartTime/map.js
* src/couchapp/AsyncTransfer/views/ftscp_all/map.js
* src/couchapp/AsyncTransfer/views/getFilesToRetry/map.js
* src/couchapp/AsyncTransfer/views/get_acquired/map.js
* src/couchapp/AsyncTransfer/views/sites/map.js
* src/couchapp/DBSPublisher/views/cache_area/map.js
* src/couchapp/UserMonitoring/views/FilesByWorkflow/map.js
* src/couchapp/UserMonitoring/views/LFNByJobId/map.js
* src/couchapp/UserMonitoring/views/LFNDoneByJobId/map.js
* src/couchapp/UserMonitoring/views/LFNFailedByJobId/map.js
* src/couchapp/UserMonitoring/views/LFNKilledByJobId/map.js
Some of these views use doc.lfn only inside if statements to check that doc.lfn exists. But since my idea is to remove doc.lfn completely, you should not use doc.lfn for anything anymore.

Other places are:
* src/couchapp/AsyncTransfer/updates/updateJobs.js
* src/python/AsyncStageOut/Plugins/Dummy.py
* src/python/AsyncStageOut/AnalyticsDaemon.py
* src/python/AsyncStageOut/TransferWorker.py where you should not do this https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/TransferWorker.py#L485 and https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/TransferWorker.py#L492 anymore
* Same comment as above for https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/ReporterWorker.py#L243 and https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/ReporterWorker.py#L250

I thought there were more places where you were using the doc.lfn and replacing store/temp <--> store. But seems not. My mistake, sorry. I got confused by things you have like 
https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/TransferWorker.py#L435 and https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/TransferWorker.py#L466-L467 where you need that because in case of direct stageout CRAB was setting doc.source_lfn = /store/user/bla (without temp), which is a different issue. In that respect, I have fixed CRAB (https://github.com/dmwm/CRABServer/pull/4893) to never remove "/temp" from the source_lfn and this will be available in September release. So whenever you want to know if a file needs to be transferred or not you should rely on the doc state, and not on the source_lfn having or not "temp" in it. I added a comment for example in the PublisherWorker about that (https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/PublisherWorker.py#L205-L213 -notice the comment refers to CRAB 3.3.1508 because I thought there was going to be an August release-). So once all documents in Couch will have source_lfn without "temp" (in October or November) you should be able to remove things like https://github.com/dmwm/AsyncStageout/blob/1.0.3pre14/src/python/AsyncStageOut/TransferWorker.py#L526-L528 (and similarly in src/python/AsyncStageOut/ReporterWorker.py and src/python/AsyncStageOut/PublisherWorker.py). But as said, this is a different issue, which I can open if you want.

In summary, if you think the places where doc.lfn is still used are irrelevant for the functionality of ASO, then you can close this use and maybe open another one for just "cleaning up the few places where doc.lfn is still there". And then the other issue for adapting ASO to the new format of doc.source_lfn (which should always have "temp" in it) if you want.

**Comment by dciangot** [comment]
done. please close it.

@mmascher @belforte
